okhttp/src/main/java/okhttp3/internal/cache/CacheInterceptor.kt

```diff
@@ -0,0 +1,267 @@
+import okio.Timeout
+import java.net.HttpURLConnection.HTTP_NOT_MODIFIED
+import java.util.concurrent.TimeUnit.MILLISECONDS
+import okhttp3.internal.Util.closeQuietly
+import okhttp3.internal.Util.discard
+import okhttp3.internal.addHeaderLenient
+import okio.buffer
+import java.net.HttpURLConnection.HTTP_GATEWAY_TIMEOUT
+
+/** Serves requests from the cache and writes responses to the cache.  */
+class CacheInterceptor(internal val cache: InternalCache?) : Interceptor {
+
+  @Throws(IOException::class)
+  override fun intercept(chain: Interceptor.Chain): Response {
+    val cacheCandidate = cache?.get(chain.request())
+
+    val now = System.currentTimeMillis()
+
+    val strategy = CacheStrategy.Factory(now, chain.request(), cacheCandidate).get()
+    val networkRequest = strategy.networkRequest
+    val cacheResponse = strategy.cacheResponse
+
+    if (cache != null) {
+      cache.trackResponse(strategy)
+    }
+
+    if (cacheCandidate != null && cacheResponse == null) {
+      // The cache candidate wasn't applicable. Close it.
+      closeQuietly(cacheCandidate.body())
+    }
+
+    // If we're forbidden from using the network and the cache is insufficient, fail.
+    if (networkRequest == null && cacheResponse == null) {
+      return Response.Builder()
+          .request(chain.request())
+          .protocol(Protocol.HTTP_1_1)
+          .code(HTTP_GATEWAY_TIMEOUT)
+          .message("Unsatisfiable Request (only-if-cached)")
+          .body(Util.EMPTY_RESPONSE)
+          .sentRequestAtMillis(-1L)
+          .receivedResponseAtMillis(System.currentTimeMillis())
+          .build()
+    }
+
+    // If we don't need the network, we're done.
+    if (networkRequest == null) {
+      return cacheResponse!!.newBuilder()
+          .cacheResponse(stripBody(cacheResponse))
+          .build()
+    }
+
+    var networkResponse: Response? = null
+    try {
+      networkResponse = chain.proceed(networkRequest)
+    } finally {
+      // If we're crashing on I/O or otherwise, don't leak the cache body.
+      if (networkResponse == null && cacheCandidate != null) {
+        closeQuietly(cacheCandidate.body())
+      }
+    }
+
+    // If we have a cache response too, then we're doing a conditional get.
+    if (cacheResponse != null) {
+      if (networkResponse?.code() == HTTP_NOT_MODIFIED) {
+        val response = cacheResponse.newBuilder()
+            .headers(combine(cacheResponse.headers(), networkResponse.headers()))
+            .sentRequestAtMillis(networkResponse.sentRequestAtMillis())
+            .receivedResponseAtMillis(networkResponse.receivedResponseAtMillis())
+            .cacheResponse(stripBody(cacheResponse))
+            .networkResponse(stripBody(networkResponse))
+            .build()
+
+        networkResponse.body()?.close()
+
+        // Update the cache after combining headers but before stripping the
+        // Content-Encoding header (as performed by initContentStream()).
+        if (cache != null) {
```

Comment:
here we know the cache is non-null, the null check should be replaced with the `!!` operator

Reply:
Fixed